### PR TITLE
override purchaseCoins for DV Chain and add sandbox option.

### PR DESCRIPTION
**Pull request by ztnark**
purchasing coins through DV Chain does not currently work because there is no account service for DV Chain. This PR overrides purchaseCoins without the need for the account service.

### CODE REVIEW

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java [hunk 1]
```diff
@@ -82,8 +95,37 @@ public BigDecimal getCryptoBalance(String cryptoCurrency){
         return BigDecimal.ZERO;
     }
 
+    @Override
+    public BigDecimal getFiatBalance(String fiatCurrency){
+        return BigDecimal.valueOf(10000000);
```
Comment (rpanak-generalbytes) [inline server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java]: Add a check of fiatCurrency parameter against getFiatCurrencies(). If it's not there, return null.

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java [hunk 2]
```diff
@@ -82,8 +95,37 @@ public BigDecimal getCryptoBalance(String cryptoCurrency){
...
+    }
+
     @Override
     public String getDepositAddress(String cryptoCurrency){
         return null;
     }
+
+    @Override
+    public String purchaseCoins(BigDecimal amount, String cryptoCurrency, String fiatCurrencyToUse, String description) {
```
Comment (rpanak-generalbytes) [inline server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java]: Where does the purchased crypto end? Do you have to have a wallet address registered with the exchange?

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/BitcoinExtension.java [hunk 3]
```diff
@@ -106,7 +106,8 @@ public IExchange createExchange(String paramString) //(Bitstamp is in built-in e
             } else if ("dvchain".equalsIgnoreCase(prefix)) {
                 String preferredFiatCurrency = FiatCurrency.USD.getCode();
                 String apiSecret = paramTokenizer.nextToken();
-                return new DVChainExchange(apiSecret, preferredFiatCurrency);
+                boolean useSandbox = Boolean.parseBoolean(paramTokenizer.nextToken());
```
[the remark this answers is not in the document]
Comment (ztnark) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/BitcoinExtension.java]: Will do

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java (hunk 1 above)
Comment (ztnark) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java]: Will do

File: server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java (hunk 2 above)
Comment (ztnark) [reply server_extensions_extra/src/main/java/com/generalbytes/batm/server/extensions/extra/bitcoin/exchanges/dvchain/DVChainExchange.java]: DV Chain manually settles up with clients daily.

**Comment by ztnark** [comment]
DV Chain manually sends coins to its clients, and doesn't really have a concept of a "balance". The plan is for this extension to enable strategy 4  "Send coins to a customer from the Hot Wallet. After that, purchase coins on the Exchange and leave them there." Is it possible to test against my own build in CAS?

**Comment by rpanak-generalbytes** [comment]
As discussed over email, you can build your own jar for server_extensions_extra and put it into your master server installation (replacing the old one). You can use batm-manage script backup command to backup your installation (along with the current DB state) before you start making changes. You can then use batm-manage restore to restore the backup you created previously (beware you'll lose any DB changes made since then!).

For the exchange strategy - that is the crucial part. The implementation depends on the strategy you want to use. If you're aiming for Strategy 4, `purchaseCoins` is probably the right place for the code (so I take back my suggestion to move the code to `sendCoins`).

Hopefully this gives you enough information to continue.

**Comment by meMuszr** [comment]
Is there going to be extensibility to implement custom strategies -- in specifically, looking to utilize strategy 4 till a limit is reached or a recurring time limit and transfer to hot wallet (reduce fees, etc.)

**Comment by rpanak-generalbytes** [comment]
@meMuszr No plans for that right now

**Comment by rpanak-generalbytes** [comment]
Closing because of no activity. Feel free to reopen if needed.

**Pull request closed without merging**